driver/utils/unicode_converter.h

```diff
@@ -0,0 +1,540 @@
+
+            const auto target_bytes_appended = target - target_prev;
+
+            if (target_bytes_appended % sizeof(CharType) != 0)
+                throw std::runtime_error("unable to convert encoding: ucnv_fromUnicode() failed to write destination buffer to symbol boundary");
+
+            target_symbols_written += target_bytes_appended / sizeof(CharType);
+
+            while (
+                trim_encoded_signature &&
+                !encoded_signature_trimmed &&
+                (target_symbols_written * sizeof(CharType)) >= encoded_signatures_to_trim_max_size_
+            ) {
+                encoded.resize(target_symbols_written); // Shrink to avoid memmove'ing unused bytes.
+                const auto consumed_symbols = consumeEncodedSignatureInPlace(encoded); // encoded is not shorter than encoded_signatures_to_trim_max_size_, so even 0 is valid.
+                target_symbols_written -= consumed_symbols;
+                target -= consumed_symbols * sizeof(CharType);
+                encoded_signature_trimmed = (consumed_symbols == 0); // This means we may try several times, and stop only when all possible duplicate leading signatures are consumed.
+            }
+
+            if (error_code == U_BUFFER_OVERFLOW_ERROR) {
+                const std::size_t source_pending_symbols = source_end - source;
+                const std::size_t target_new_size = target_symbols_written + source_pending_symbols;
+                const std::size_t final_new_size = std::max(target_new_size + extra_size_reserve, encoded.size() + min_size_increment);
+
+                resize_without_initialization(encoded, final_new_size);
+
+                target = const_cast<char *>(reinterpret_cast<const char *>(encoded.c_str() + target_symbols_written));
+                target_end = reinterpret_cast<const char *>(encoded.c_str() + encoded.size());
+            }
+            else if (U_FAILURE(error_code)) {
+                throw std::runtime_error(u_errorName(error_code));
+            }
+            else {
+                encoded.resize(target_symbols_written);
+                break;
+            }
+        }
+    }
+    catch (...) {
+        ucnv_resetFromUnicode(converter_);
+        encoded.clear();
+        throw;
+    }
+}
+
+template <typename SourceCharType, typename DestinationCharType, typename PivotCharType>
+inline void convertEncoding(
+    UnicodeConverter & src_converter, const std::basic_string_view<SourceCharType> & src,
+    std::basic_string<PivotCharType> & pivot,
+    UnicodeConverter & dest_converter, std::basic_string<DestinationCharType> & dest,
+    const bool ensure_src_signature,
+    const bool trim_dest_signature
+) {
+#if defined(WORKAROUND_ICU_USE_EXPLICIT_PIVOTING)
+    src_converter.convertToPivot(src, pivot, ensure_src_signature, false);
+    dest_converter.convertFromPivot(make_string_view(pivot), dest, true, trim_dest_signature);
+#else
+    dest.clear();
```

Comment:
Please add a simplified plan/high-level overview of what is happening here in overly simplified pseudocode or just plain English.
Plus a note why you need a separate function and how it is different from `convertToPivot` and `convertFromPivot`.

Reply:
Done.